discovery/hetzner/hetzner.go

```diff
@@ -0,0 +1,173 @@
+	Endpoint string `yaml:"endpoint"`
+}
+
+// UnmarshalYAML implements the yaml.Unmarshaler interface.
+func (c *SDConfig) UnmarshalYAML(unmarshal func(interface{}) error) error {
+	*c = DefaultSDConfig
+	type plain SDConfig
+	err := unmarshal((*plain)(c))
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// Discovery periodically performs Hetzner Cloud requests. It implements
+// the Discoverer interface.
+type Discovery struct {
+	*refresh.Discovery
+	client  *hcloud.Client
+	port    int
+	network *hcloud.Network
+}
+
+// NewDiscovery returns a new Discovery which periodically refreshes its targets.
+func NewDiscovery(conf *SDConfig, logger log.Logger) (*Discovery, error) {
+	d := &Discovery{
+		port: conf.Port,
+	}
+
+	rt, err := config_util.NewRoundTripperFromConfig(conf.HTTPClientConfig, "hetzner_sd", false)
+	if err != nil {
+		return nil, err
+	}
+	opts := []hcloud.ClientOption{
+		hcloud.WithToken(string(conf.HTTPClientConfig.BearerToken)),
+		hcloud.WithApplication("Prometheus", version.Version),
+		hcloud.WithHTTPClient(&http.Client{
+			Transport: rt,
+			Timeout:   5 * time.Duration(conf.RefreshInterval),
+		}),
+		hcloud.WithPollInterval(time.Duration(conf.RefreshInterval)),
+		hcloud.WithEndpoint(conf.HcloudSDConfig.Endpoint),
+	}
+	d.client = hcloud.NewClient(
+		opts...,
+	)
+	if conf.HcloudSDConfig.Network != "" {
+		d.network, _, err = d.client.Network.Get(context.Background(), conf.HcloudSDConfig.Network)
+		if err != nil {
+			return nil, err
+		}
+		if d.network == nil {
+			return nil, fmt.Errorf("error finding requested network: %s", conf.HcloudSDConfig.Network)
+		}
+	}
+	d.Discovery = refresh.NewDiscovery(
+		logger,
+		"hetzner",
+		time.Duration(conf.RefreshInterval),
+		d.refresh,
+	)
+	return d, nil
+}
+func (d *Discovery) refresh(ctx context.Context) ([]*targetgroup.Group, error) {
+	servers, err := d.client.Server.All(ctx)
+	if err != nil {
+		return nil, err
+	}
+	var targets []model.LabelSet
+	for _, server := range servers {
+		labels := model.LabelSet{
+			hcloudLabelID:         model.LabelValue(fmt.Sprintf("%d", server.ID)),
+			hcloudLabelName:       model.LabelValue(server.Name),
+			hcloudLabelLocation:   model.LabelValue(server.Datacenter.Location.Name),
+			hcloudLabelDatacenter: model.LabelValue(server.Datacenter.Name),
+			hcloudLabelPublicIPv4: model.LabelValue(server.PublicNet.IPv4.IP.String()),
+			hcloudLabelType:       model.LabelValue(server.ServerType.Name),
+			hcloudLabelCPUCores:   model.LabelValue(fmt.Sprintf("%d", server.ServerType.Cores)),
+			hcloudLabelCPUType:    model.LabelValue(server.ServerType.CPUType),
+			hcloudLabelMemory:     model.LabelValue(fmt.Sprintf("%v", server.ServerType.Memory)),
+			hcloudLabelDisk:       model.LabelValue(fmt.Sprintf("%d", server.ServerType.Disk)),
+
+			model.AddressLabel: model.LabelValue(net.JoinHostPort(server.PublicNet.IPv4.IP.String(), strconv.FormatUint(uint64(d.port), 10))),
+		}
+
+		if server.Image.Name == "" {
```

Comment:
Don't munge, put out the raw metadata and leave it up to relabelling.

Reply:
The name is only set for images of type system so it could be empty when you have a snapshot or a backup in that case the description is set. Should i add a new label "hetzner_hcloud_image_description" ?

Reply:
If it's empty, it's empty.